# Non-Linear SVM - Email Spam Classifier


In this section, we'll build a non-linear SVM classifier to classify emails and compare the performance with the linear SVM model.

The dataset can be downloaded here: https://archive.ics.uci.edu/ml/datasets/spambase

To reiterate, the performance of the linear model was as follows:
- accuracy 0.93
- precision 0.92
- recall 0.89

In this section, we will build a non-linear model (using non-linear kernels) and then find the optimal hyperparameters (the choice of kernel, C, gamma).

In [1]:
import pandas as pd
import numpy as np
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import scale

**Loading Data**

In [5]:
email_rec = pd.read_csv("Spam (1).csv",  sep = ',')

## Data Preparation

In [6]:
# splitting into X and y
X = email_rec.drop("spam", axis = 1)
y = email_rec.spam.values.astype(int)

In [7]:
# scaling the features
X_scaled = scale(X)

# train test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, random_state = 4)

## Model Building

In [8]:
# using rbf kernel, C=1, default value of gamma

model = SVC(C = 1, kernel='rbf')
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

## Model Evaluation Metrics

In [9]:
# confusion matrix
confusion_matrix(y_true=y_test, y_pred=y_pred)

array([[811,  38],
       [ 61, 471]])

In [10]:
# accuracy
print("accuracy", metrics.accuracy_score(y_test, y_pred))

# precision
print("precision", metrics.precision_score(y_test, y_pred))

# recall/sensitivity
print("recall", metrics.recall_score(y_test, y_pred))

accuracy 0.928312816799
precision 0.925343811395
recall 0.885338345865


## Hyperparameter Tuning 

Now, we have multiple hyperparameters to optimise - 
- The choice of kernel (linear, rbf etc.)
- C
- gamma

We'll use the ```GridSearchCV()``` method to tune the hyperparameters. 

## Grid Search to Find Optimal Hyperparameters

Let's first use the RBF kernel to find the optimal C and gamma (we can consider the kernel as a hyperparameter as well, though training the model will take an exorbitant amount of time). 

In [11]:
# creating a KFold object with 5 splits 
folds = KFold(n_splits = 5, shuffle = True, random_state = 4)

# specify range of hyperparameters
# Set the parameters by cross-validation
hyper_params = [ {'gamma': [1e-2, 1e-3, 1e-4],
                     'C': [1, 10, 100, 1000]}]


# specify model
model = SVC(kernel="rbf")

# set up GridSearchCV()
model_cv = GridSearchCV(estimator = model, 
                        param_grid = hyper_params, 
                        scoring= 'accuracy', 
                        cv = folds, 
                        verbose = 1,
                        return_train_score=True)      

# fit the model
model_cv.fit(X_train, y_train)                  


Fitting 5 folds for each of 12 candidates, totalling 60 fits


[Parallel(n_jobs=1)]: Done  60 out of  60 | elapsed:   50.3s finished


GridSearchCV(cv=KFold(n_splits=5, random_state=4, shuffle=True),
       error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid=[{'gamma': [0.01, 0.001, 0.0001], 'C': [1, 10, 100, 1000]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring='accuracy', verbose=1)

In [12]:
# cv results
cv_results = pd.DataFrame(model_cv.cv_results_)
cv_results

mean_fit_time  mean_score_time  mean_test_score  mean_train_score param_C  \
0        0.517946         0.088059         0.929814          0.941304       1   
1        0.473316         0.092060         0.904037          0.906522       1   
2        0.805738         0.166712         0.786025          0.786957       1   
3        0.278786         0.046831         0.933230          0.964752      10   
4        0.333622         0.053636         0.928261          0.934472      10   
5        0.451701         0.093662         0.902174          0.905745      10   
6        0.426285         0.049633         0.931677          0.981910     100   
7        0.331622         0.045830         0.933851          0.946817     100   
8        0.340226         0.059240         0.927019          0.931910     100   
9        0.523549         0.032022         0.918323          0.992857    1000   
10       0.948431         0.074650         0.933851          0.965761    1000   
11       0.412275         0.056638         0.929193          0.939441    1000   

   param_gamma                        params  rank_test_score  \
0         0.01       {'C': 1, 'gamma': 0.01}                5   
1        0.001      {'C': 1, 'gamma': 0.001}               10   
2       0.0001     {'C': 1, 'gamma': 0.0001}               12   
3         0.01      {'C': 10, 'gamma': 0.01}                3   
4        0.001     {'C': 10, 'gamma': 0.001}                7   
5       0.0001    {'C': 10, 'gamma': 0.0001}               11   
6         0.01     {'C': 100, 'gamma': 0.01}                4   
7        0.001    {'C': 100, 'gamma': 0.001}                1   
8       0.0001   {'C': 100, 'gamma': 0.0001}                8   
9         0.01    {'C': 1000, 'gamma': 0.01}                9   
10       0.001   {'C': 1000, 'gamma': 0.001}                1   
11      0.0001  {'C': 1000, 'gamma': 0.0001}                6   

    split0_test_score  split0_train_score       ...         split2_test_score  \
0            0.917702            0.943323       ...                  0.922360   
1            0.886646            0.910326       ...                  0.899068   
2            0.770186            0.789208       ...                  0.791925   
3            0.909938            0.966227       ...                  0.934783   
4            0.917702            0.937112       ...                  0.916149   
5            0.883540            0.909938       ...                  0.899068   
6            0.913043            0.982531       ...                  0.934783   
7            0.923913            0.950311       ...                  0.925466   
8            0.919255            0.934006       ...                  0.917702   
9            0.908385            0.993789       ...                  0.920807   
10           0.919255            0.966615       ...                  0.930124   
11           0.920807            0.940606       ...                  0.925466   

    split2_train_score  split3_test_score  split3_train_score  \
0             0.945264           0.931677            0.937112   
1             0.908773           0.895963            0.906056   
2             0.785326           0.765528            0.791925   
3             0.967003           0.934783            0.961568   
4             0.936335           0.933230            0.935171   
5             0.908773           0.892857            0.905280   
6             0.982531           0.930124            0.982143   
7             0.948758           0.933230            0.945652   
8             0.934006           0.928571            0.930901   
9             0.992624           0.916149            0.993012   
10            0.966615           0.933230            0.963121   
11            0.940994           0.928571            0.937112   

    split4_test_score  split4_train_score  std_fit_time  std_score_time  \
0            0.937888            0.939829      0.068217        0.010282   
1            0.919255            0.904115      0.02661

In [13]:
# converting C to numeric type for plotting on x-axis
cv_results['param_C'] = cv_results['param_C'].astype('int')

# # plotting
plt.figure(figsize=(16,6))

# subplot 1/3
plt.subplot(131)
gamma_01 = cv_results[cv_results['param_gamma']==0.01]

plt.plot(gamma_01["param_C"], gamma_01["mean_test_score"])
plt.plot(gamma_01["param_C"], gamma_01["mean_train_score"])
plt.xlabel('C')
plt.ylabel('Accuracy')
plt.title("Gamma=0.01")
plt.ylim([0.80, 1])
plt.legend(['test accuracy', 'train accuracy'], loc='upper left')
plt.xscale('log')

# subplot 2/3
plt.subplot(132)
gamma_001 = cv_results[cv_results['param_gamma']==0.001]

plt.plot(gamma_001["param_C"], gamma_001["mean_test_score"])
plt.plot(gamma_001["param_C"], gamma_001["mean_train_score"])
plt.xlabel('C')
plt.ylabel('Accuracy')
plt.title("Gamma=0.001")
plt.ylim([0.80, 1])
plt.legend(['test accuracy', 'train accuracy'], loc='upper left')
plt.xscale('log')


# subplot 3/3
plt.subplot(133)
gamma_0001 = cv_results[cv_results['param_gamma']==0.0001]

plt.plot(gamma_0001["param_C"], gamma_0001["mean_test_score"])
plt.plot(gamma_0001["param_C"], gamma_0001["mean_train_score"])
plt.xlabel('C')
plt.ylabel('Accuracy')
plt.title("Gamma=0.0001")
plt.ylim([0.80, 1])
plt.legend(['test accuracy', 'train accuracy'], loc='upper left')
plt.xscale('log')


This plot reveals some interesting insights:
- **High values of gamma** lead to **overfitting** (especially at high values of C); note that the training accuracy at gamma=0.01 and C=1000 reaches almost 99% 
- The **training score increases with higher gamma**, though the **test scores are comparable** (at sufficiently high cost, i.e. C > 10)
- The least amount of overfitting (i.e. difference between train and test accuracy) occurs at low gamma, i.e. a quite *simple non-linear model*


In [14]:
# printing the optimal accuracy score and hyperparameters
best_score = model_cv.best_score_
best_hyperparams = model_cv.best_params_

print("The best test score is {0} corresponding to hyperparameters {1}".format(best_score, best_hyperparams))

The best test score is 0.9338509316770186 corresponding to hyperparameters {'C': 100, 'gamma': 0.001}


Though sklearn suggests the optimal scores mentioned above (gamma=0.001, C=100), one could argue that it is better to choose a simpler, more non-linear model with gamma=0.0001. This is because the optimal values mentioned here are calculated based on the average test accuracy (but not considering subjective parameters such as model complexity).

We can achieve comparable average test accuracy (~92.5%) with gamma=0.0001 as well, though we'll have to increase the cost C for that. So to achieve high accuracy, there's a tradeoff between:
- High gamma (i.e. high non-linearity) and average value of C
- Low gamma (i.e. less non-linearity) and high value of C

We argue that the model will be simpler if it has as less non-linearity as possible, so we choose gamma=0.0001 and a high C=100.

### Building and Evaluating the Final Model

Let's now build and evaluate the final model, i.e. the model with highest test accuracy.

In [15]:
# specify optimal hyperparameters
best_params = {"C": 100, "gamma": 0.0001, "kernel":"rbf"}

# model
model = SVC(C=100, gamma=0.0001, kernel="rbf")

model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# metrics
print(metrics.confusion_matrix(y_test, y_pred), "\n")
print("accuracy", metrics.accuracy_score(y_test, y_pred))
print("precision", metrics.precision_score(y_test, y_pred))
print("sensitivity/recall", metrics.recall_score(y_test, y_pred))

[[810  39]
 [ 60 472]] 

accuracy 0.928312816799
precision 0.923679060665
sensitivity/recall 0.887218045113


## Conclusion

The accuracy achieved using a non-linear kernel is comparable to that of a linear one. Thus, it turns out that for this problem, **you do not really need a non-linear kernel**.